[cdac] Physical contract descriptor spec

lambdageek opened the pull request:
Building on #100253 , describe an in-memory representation of the toplevel contract descriptor, comprisied of:
* some target architecture properties
* a data descriptor
* a collection of compatible contracts

Contributes to #99298 
Fixes https://github.com/dotnet/runtime/issues/99299

lambdageek committed:
[cdac] Physical contract descriptor spec

davidwrighton commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,63 @@
+target platform.  That is, on platforms where such symbols typicall have an `_` prepended, this
+symbol should be exported as `_DotNetRuntimeContractDescriptor`)
+
+**FIXME** What about scenarios such as a NativeAOT library hosted inside a native process?  What if
```
My suspicion is that we'll want to be a bit weird with this. On Windows a model where dlls are naturally a symbol boundary, it makes sense to have a symbol per dll, but on Unix platforms this is less reasonable. Instead, there I would spec that there is a single global symbol  that represents a pointer to a linked list of `DotNetRuntimeContractDescriptor`  ... Call it `_DotNetRuntimeContractDescriptorList`, and define it as a "weak" symbol, so that multiple SOs in the process can define it, and the dynamic linker will pick one to "win". Then, on startup, every runtime in the process will add its contract descriptor to the linked list. Then it becomes a problem for the diagnostics tool to figure out which one it wants to debug. We may wish to add a "name" string to the contract data descriptor, just so that we can have some way to express to the end diagnostic tool user that data X is from "CoreCLR v10.38.437.4" and data "Y" is from "NativeAOTLibraryFoo". This implies that we should change your definition of the `DontNetRuntimeContractDescriptor` to include a field for a linked list, and define that this entries in the linked list with a different magic number should be "ignored" (this allows for versioning of the descriptor while not breaking old diagnostic tooling). Logically we can do the same thing on Windows, if we want to support linking a NativeAOT runtime into coreclr or something. We would just end up with a need to scan all dlls modules for their lists and concatenate the lists together.

lambdageek replied:
Done. Hopefully the unixy bits aren't science fiction

davidwrighton reviewed (commented):
I think this is a good direction, but I have a few comments.

davidwrighton commented on docs/design/datacontracts/contract-descriptor.md:
````diff
@@ -0,0 +1,63 @@
+The `compatible_contracts` are a json string giving the [compatible contracts](./datacontracts_design.md#Compatible_Contract).  The compatible contracts are given as a json array where each element is a dictionary.  The dictionary will have a `c` key giving the name of the compatible contract as a string, and a `v` key giving the contract version as an integer.  For example:
+
+``` jsonc
+[{"c":"Thread","v":1},{"c":"GCHandle","v":1},...]
````
I like that. It avoids the odd string parsing proposal I had, and is even more compact.

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,63 @@
+    uint32_t aux_data_count;
+    char *data_descriptor;
+    uint64_t *aux_data;
+    char *compatible_contracts;
```
Does it need to be a separate json? I think both `data_descriptor` and `compatible_contracts` can be together in one json

lambdageek replied:
Merged the two json blobs together

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,63 @@
+    uint64_t magic;
+    uint32_t flags;
+    uint32_t aux_data_count;
+    char *data_descriptor;
```
It may be convenient to have the size of the json by bytes in the header too. It makes it easy to read it from a file or out of proc.

lambdageek committed:
Add "contracts" to the data descriptor; spec Unix-y weak symbol shennanigans

lambdageek committed:
add a runtime name

lambdageek committed:
reword and constness

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,151 @@
+void
+install_descriptor(const char *runtime_name)
+{
+    DescPtr descriptor = malloc(sizeof(struct DotNetRuntimeContractDescriptorList));
```
Can this be a local static variable instead?

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,151 @@
+    do
+    {
+        descriptor->next_runtime = prev;
+    } while (!atomic_compare_exchange_weak(&DotNetRuntimeConctractDescriptor, &prev, descriptor));
```
```suggestion
    } while (!atomic_compare_exchange_weak(&DotNetRuntimeContractDescriptor, &prev, descriptor));
```

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,151 @@
+{
+    DescPtr descriptor = malloc(sizeof(struct DotNetRuntimeContractDescriptorList));
+    assert (descriptor != NULL);
+    descriptor->descriptor = g_private_descriptor; // copy the constant values
```
Can this a be pointer instead of a copy?

lambdageek replied:
I may need to revisit this.  The issue is that we don't know the pointer size and endianness until we read off the magic and the flags.  

On the other hand that's somewhat untrue - since we're relying on the diagnostic tool's ability to find the addresses of symbols in the target process at some level the tool must already know what architecture it's looking at.

lambdageek replied:
Actually this whole list mechanism just doesn't work without knowing up front the target pointer size and endianness.

Is there really a scenario where the diagnostic tooling doesn't know the architecture of the target process?  Do minidumps have that info?

mikem8361 replied:
SOS can deal with multiple runtimes in a process.  The `!runtimes` is the command to list and switch between them.  SOS/CLRMD (SOS uses CLRMD's runtime enumeration) currently does include any Native AOT modules. The internal DCR based SOS extensions does this Native AOT runtime enumeration. 

When there is enough cDAC support for Native AOT, we can change the SOS/CLRMD to include them.

lambdageek replied:
Removed the `DotNetRuntimeContractDescriptorList`.  clarified that for multi-runtime scenarios the diagnostic tooling should look in each loaded module for the contract descriptor

lambdageek committed:
fixup example

lambdageek committed:
markdownlint

lambdageek committed:
don't use a pointer

the issue is that we can't dereference pointers until we know the
target endianness and pointer size.  which we can only discover by
reading the actual descriptor magic and flags

lambdageek committed:
flags not size_and_flags

lambdageek committed:
remove DotNetRuntimeContractDescriptorList - one runtime per module

if there are multiple hosted runtimes, diagnostic tooling should look
in each loaded module to discover the contract descriptor

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,101 @@
+
+To aid in the discovery the contract descriptor should be exported by the module hosting the .NET
+runtime with the name `DotNetRuntimeContractDescriptor`. (Using the C symbol conventions of the
+target platform.  That is, on platforms where such symbols typically have an `_` prepended, this
```
What are those platforms?

FWIW, the existing `DotNetRuntimeDebugHeader` seems to be exported without `_` everywhere: https://github.com/dotnet/runtime/blob/ab2bd48741b54baa9ac416a3da56efe915826085/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L245-L246

jkotas commented on docs/design/datacontracts/contract-descriptor.md:
```diff
@@ -0,0 +1,101 @@
+If `ptrSize` is 0, the architecture is 64-bit.  If it is 1, the architecture is 32-bit.  The
+reserved bits should be written as zero.  Diagnostic tooling may ignore non-zero reserved bits.
+
+The `descriptor` is a pointer to a JSON string described in [data descriptor physical layout](./data_descriptor.md#Physical_JSON_descriptor).  The total length (including nul terminator character) is given by `descriptor_size`.
```
Nit: I would omit the terminating null in the `descriptor_size`.

The data contract reader cannot assume that the descriptor is null terminated. It needs to be robust against malicious input. Including the terminating null does not seem to be beneficial, it just adds extra logic to the reader to verify that there is a null at the end and strip it.

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
Review feedback

- put the aux data and descriptor sizes closer to the pointers

- Don't include trailing nul `descriptor_size`.  Clarify it is
counting bytes and that `descriptor` is in UTF-8

- Simplify `DotNetRuntimeContractDescriptor` naming discussion
